linera-rpc/src/client.rs

```diff
@@ -153,4 +154,16 @@ impl ValidatorNode for Client {
             Client::Simple(simple_client) => simple_client.download_blob(blob_id).await?,
         })
     }
+
+    async fn download_certificate_value(
+        &mut self,
+        hash: CryptoHash,
+    ) -> Result<CertificateValue, NodeError> {
```

Comment:
We could use `HashedCertificateValue` everywhere: Its `Serialize` and `Deserialize` are implemented such that the hash is not explicitly serialized, but computed on deserialization.

Maybe hashing on deserialization is controversial, but then we should change that everywhere: Let's try to be consistent between e.g. `HashedBlob` and `HashedCertificateValue`, and let's find a way to:
* Make sure the hash is never deserialized from an untrusted source.
* Avoid computing the hash more often than once.

Reply:
After discussing with @ma2bd, let's move in the following direction for both blobs and certificate values:

* `Serialize` and `Deserialize` are only implemented for the non-hashed versions.
* There are explicit conversion functions:
  - `with_hash` computes the hash, converting the non-hashed to the hashed type.
  - `with_hash_unchecked` adds a given hash, without computing it, converting the non-hashed to the hashed type.
  - `forget_hash` converts the hashed to the non-hashed type.
* At public boundaries, i.e. any communication between the client and a node (server or proxy), the non-hashed version is sent, and the hash is computed on deserialization.
* In storage, the hash is stored separately (usually as the key), and after deserialization it is added using `with_hash_unchecked`.

Reply:
I created a ticket for this: https://github.com/linera-io/linera-protocol/issues/2136